src/opencmiss/zincwidgets/graphicseditorwidget.py

```diff
@@ -716,3 +1052,53 @@ def samplingModeChanged(self, index):
             if samplingattributes.isValid():
                 samplingMode = self._ui.sampling_mode_chooser.getEnum()
                 samplingattributes.setElementPointSamplingMode(samplingMode)
+                self._sampleLocationDisplay()
+
+    def _sampleLocationDisplay(self):
+        """
+        Show current sample location
+        """
+        self._ui.sample_location_lineedit.setEnabled(False)
+        if self._graphics:
+            samplingattributes = self._graphics.getGraphicssamplingattributes()
+            if samplingattributes.isValid() and samplingattributes.getElementPointSamplingMode() == Element.POINT_SAMPLING_MODE_SET_LOCATION:
+                self._ui.sample_location_lineedit.setEnabled(True)
+                _, scaleFactors = samplingattributes.getLocation(3)
+                self._displayVector(self._ui.sample_location_lineedit, scaleFactors)
+                return
+        self._ui.sample_location_lineedit.setText('')
+
+    def sampleLocationEntered(self):
+        """
+        Set sample location from text in widget
+        """
+        try:
+            sampleLocation = self._parseVector(self._ui.sample_location_lineedit)
+            samplingattributes = self._graphics.getGraphicssamplingattributes()
+            if samplingattributes.setLocation(sampleLocation) != ZINC_OK:
+                raise
+        except:
+            print("Invalid sample location")
+        self._sampleLocationDisplay()
+
+    def _densityFieldDisplay(self):
+        """
+        Show the current state of the density field combo box
+        """
+        if self._graphics:
+            samplingattributes = self._graphics.getGraphicssamplingattributes()
+            if samplingattributes.isValid():
+                densityField = samplingattributes.getDensityField()
+                self._ui.density_field_chooser.setField(densityField)
+                return
+
+    def densityFieldChanged(self):
+        """
+        Density field combo box changed
+        """
+        if self._graphics:
+            samplingattributes = self._graphics.getGraphicssamplingattributes()
+            if samplingattributes.isValid():
+                densityField = self._ui.density_field_chooser.getField()
+                samplingattributes.setDensityField(densityField)
+                self._densityFieldDisplay()
```

Comment:
Add carriage return CR -- don't end a file on a line with code (PEP 8 considers it a problem).